src/hooks/useWarnUserForUnsavedChanges.ts

```diff
@@ -0,0 +1,42 @@
+import { useEffect } from "react"
+export default function useWarnUserForUnsavedChanges({
+  hasUnsavedChanges,
+}: {
+  hasUnsavedChanges: boolean
+}) {
+  useEffect(() => {
+    const handleBeforeUnload = (event: BeforeUnloadEvent) => {
+      if (hasUnsavedChanges) {
+        event.preventDefault()
+        event.returnValue = "" // Shows the confirmation dialog on reload or close if there are unsaved changes
+      }
+    }
+
+    const handlePopState = (event: PopStateEvent) => {
+      if (hasUnsavedChanges) {
+        const userConfirmed = window.confirm(
+          "Are you sure you want to go back? Unsaved changes may be lost.",
+        )
+        if (!userConfirmed) {
+          // Push the state again to stay on the same page
+          window.history.pushState(null, "", window.location.href)
```

Comment:
do we have to push state to stay on the same page? Can we just not do anything and block the action?

Reply:
state is necessary here this solution works by storing the current url in the state and if user clicks on cancel instead of sending it to previous page it send it to the current page

Reply:
What would happen if you removed the pushState?

Reply:
if i understand your code correctly, `preventDefault` stops the page from changing, so I'm not sure why you need to pushState

Reply:
It basically stores the current url (editor page) and only allows when cancelled. Is there a problem with storing the data in the state?

Reply:
Generally we only introduce state when absolutely necessary. If it's not necessary, we shouldn't do it.

I'm not convinced it's necessary. You only need pushState to change to a new page, and we're cancelling a navigation, so there should never be a need to pushState

Reply:
the thing is preventDefault will work for case like reloading but in case of user going back it will not work

Reply:
we can try to do this with pop state functionality but it doesn't block the action rather using it we can push state to restore current page

Reply:
I believe to add this functionality we will need to have states involved. However, if you have better method in mind I can make it .

Reply:
let's remove the pushStates, if we only work for reloading that's OK for now.

Reply:
as you wish but i will just comment out that part just in case you want to update

Reply:
let's just remove it, i'd rather not have ghost code, people can come back and look at this PR.

Also make sure to revert the changes to files where things were accidentally changed. I'll leave some comments that indicate where that happened. There should be no changes to `CodeEditor` (IIUC), you can run `git checkout origin/main -- path/to/file` to revert a file

Reply:
ok will do